command/agent/alloc_endpoint.go

```diff
@@ -105,43 +106,70 @@ func (s *HTTPServer) ClientAllocRequest(resp http.ResponseWriter, req *http.Requ
 }
 
 func (s *HTTPServer) ClientGCRequest(resp http.ResponseWriter, req *http.Request) (interface{}, error) {
-	if s.agent.client == nil {
-		return nil, clientNotRunning
-	}
-
-	var secret string
-	s.parseToken(req, &secret)
-
-	// Check node write permissions
-	if aclObj, err := s.agent.Client().ResolveToken(secret); err != nil {
-		return nil, err
-	} else if aclObj != nil && !aclObj.AllowNodeWrite() {
-		return nil, structs.ErrPermissionDenied
+	// Get the requested Node ID
+	requestedNode := req.URL.Query().Get("node_id")
+
+	// Build the request and parse the ACL token
+	args := structs.NodeSpecificRequest{
+		NodeID: requestedNode,
+	}
+	s.parse(resp, req, &args.QueryOptions.Region, &args.QueryOptions)
+
+	// Determine the handler to use
+	useLocalClient, useClientRPC, useServerRPC := s.rpcHandlerForNode(requestedNode)
+
+	// Make the RPC
+	var reply structs.GenericResponse
+	var rpcErr error
+	if useLocalClient {
```

Comment:
Is there a better way to dispatch this? This might be cleaner pulled into a helper or using an interface for each implementation perhaps.